Restore a node's enrollment data from the Genesis block

linked0 opened the pull request:
This PR is for restoring the enrollment data of a node from the `enrollments` field of the header of blocks. This is the first step to remove the `node_enroll_data` table in the `EnrollmentManager` class, which is to remove the `Enrollment` object from the table. The next step is to remove `next_reveal_height` from the table. Finally, the table will be removed.

(EDIT)
This PR has added two more items as follows
- Restoring the enrollment data when the `EnrollmentManager` adds an validator to the `ValidatorSet`.
- Restoring the pre-image data from blocks of the chain

Fixes #789

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -957,3 +967,68 @@ unittest
     assert(result_image.distance == 1100);
 }
+
+/// test for restroing enrollment data for a node from blocks
```
```suggestion
/// test for restoring enrollment data for a node from blocks
```

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -116,7 +118,7 @@ public class Ledger
                 this.storage.readBlock(block, this.last_block.header.height - i);
                 this.enroll_man.restoreValidators(this.last_block.header.height,
-                    block, this.utxo_set.getUTXOFinder());
+                    block, this.utxo_set.getUTXOFinder(), utxos);
```
Is this just restoring the current node's enrollments? What about other nodes' enrollments?

Geod24 replied:
They are restored already (see L551).

AndrejMitrovic replied:
L551? In this PR? It's pointing to `auto pool = new TransactionPool(":memory:");`

linked0 replied:
He means L551 of the EnrollmentManager.d

AndrejMitrovic replied:
Right, that looks ok then.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -541,13 +541,23 @@ public class EnrollmentManager
         this.validator_set.restoreValidators(last_height, block, finder);
+        foreach (const ref enroll; block.header.enrollments)
+        {
+            if ((enroll.utxo_key in node_utxos) != null)
```
```suggestion
            if (enroll.utxo_key in node_utxos)
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -541,13 +541,23 @@ public class EnrollmentManager
             last_height = the latest block height
             block = the block to update the validator set with
             finder = the delegate to find UTXOs with
+            node_utxos = the UTXOs belonging to a node
```
```suggestion
            node_utxos = the UTXOs belonging to this node
```

Geod24 commented:
We had a discussion about it with @linked0 and I still think it's the wrong approach.
The issue is relying on `enroll_key` to be filled. It's possible that it's not (in a very rare  case, but still). Instead we should simply use the node's public key, which we are sure we have.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -108,6 +108,8 @@ public class Ledger
         if (this.last_block.header.height > 0)
         {
+            UTXOSetValue[Hash] utxos =
+                this.utxo_set.getUTXOs(this.node_config.key_pair.address);
```
This assumes the node is a Validator node. A FullNode will not have a keypair.

linked0 replied:
Hmm. It can be a full node.

AndrejMitrovic replied:
As a workaround you could do:

```D
UTXOSetValue[Hash] utxos;
if (this.node_config.is_validator)
    utxos = this.utxo_set.getUTXOs(this.node_config.key_pair.address);
```

linked0 replied:
Thank you. But, It must get UTXOs using a public key although it is a full node.

linked0 replied:
> Thank you. But, It must get UTXOs using a public key although it is a full node.

I'm trying to convert the `V` of `Pair`  into `address` of `KeyPair` based on what @AndrejMitrovic has given as a clue, which is adding the `opSlice` function for the `Point` structure. But it takes some more time than expected. So I'd like to have a pair programming session for that. Thanks.

linked0 commented:
This PR can go on after the PR #823  is approved.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -541,13 +541,23 @@ public class EnrollmentManager
 
     public void restoreValidators (ulong last_height, const ref Block block,
-        scope UTXOFinder finder) @safe nothrow
+        scope UTXOFinder finder, ref UTXOSetValue[Hash] node_utxos) @safe nothrow
```
```suggestion
        scope UTXOFinder finder, const ref UTXOSetValue[Hash] node_utxos) @safe nothrow
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -541,13 +541,23 @@ public class EnrollmentManager
+        scope UTXOFinder finder, ref UTXOSetValue[Hash] node_utxos) @safe nothrow
     {
         this.validator_set.restoreValidators(last_height, block, finder);
+        foreach (const ref enroll; block.header.enrollments)
+        {
+            if (enroll.utxo_key in node_utxos)
+            {
+                this.data = enroll;
+                this.enroll_key = enroll.utxo_key;
+                break;
+            }
+        }
```
```suggestion
        if (node_utxos.length == 0)
            return;
 
        foreach (const ref enroll; block.header.enrollments)
        {
            if (enroll.utxo_key in node_utxos)
            {
                this.data = enroll;
                this.enroll_key = enroll.utxo_key;
                break;
            }
        }
```

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -108,6 +108,8 @@ public class Ledger
         // restore validator set from lastest blocks
         if (this.last_block.header.height > 0)
         {
+            UTXOSetValue[Hash] utxos =
+                this.utxo_set.getUTXOs(this.node_config.key_pair.address);
```
```suggestion
            UTXOSetValue[Hash] utxos;
            if (this.node_config.is_validator)
                utxos = this.utxo_set.getUTXOs(this.node_config.key_pair.address);
```

Because a FullNode does not contain its own enrollments, only a Validator does.

linked0 commented:
> We had a discussion about it with @linked0 and I still think it's the wrong approach.
> The issue is relying on `enroll_key` to be filled. It's possible that it's not (in a very rare case, but still). Instead we should simply use the node's public key, which we are sure we have.

We need to have white boarding session about this.

AndrejMitrovic commented:
> The issue is relying on enroll_key to be filled. It's possible that it's not

I'm not seeing the full picture here. I thought the PR **recovers** it from the blockchain?

linked0 commented:
Ready for review.
The network test will be added in another PR because the `node.ctrl.restart` which is needed for the test is being implemented now by @Geod24 .

Geod24 commented:
Let's get it through the door: https://github.com/Geod24/localrest/pull/71

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -687,13 +687,31 @@ public class EnrollmentManager
             last_height = the latest block height
             block = the block to update the validator set with
             finder = the delegate to find UTXOs with
+            node_utxos = the UTXOs belonging to this node
```
```suggestion
            self_utxos = the UTXOs belonging to this node
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -687,13 +687,31 @@ public class EnrollmentManager
 
     public void restoreValidators (ulong last_height, const ref Block block,
-        scope UTXOFinder finder) @safe nothrow
+        scope UTXOFinder finder, const ref UTXOSetValue[Hash] node_utxos)
```
```suggestion
        scope UTXOFinder finder, const ref UTXOSetValue[Hash] self_utxos)
```

That should make it slightly clearer IMO (but let me know if you disagree).

linked0 replied:
> scope UTXOFinder finder, const ref UTXOSetValue[Hash] self_utxos)

I guess it's clearer.

Geod24 reviewed (commented):
Looks nice, although I think this is a typical case where a network test is mandatory.

linked0 commented:
I'll add a test code using the `restart` function of the `LocalRest` in another commit.

linked0 commented:
Ready for review.
It doesn't have network test code yet but I added new commit for restoring pre-image information such as `cycle.index` from the chain. The new commit should be on another PR, but the code is tightly related to the commits on this PR. It makes this PR bigger, so sorry for your inconvenience.

For network test code, I need to start coding with the new `restart` function of `LocalRest`. Before that, I ask you to review my new commit.
Thanks.

linked0 commented:
The network test for this PR is being written and modified in the PR #893  . After the code gets complete, the commit of the test code will be cherry-picked into this PR.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -113,6 +113,9 @@ public class EnrollmentManager
     private PreImageCycle cycle;
 
+    /// Index of pre-image cycle restored from the chain
+    private uint restored_cycle_index;
```
No. As I said countless times, we shouldn't have a special path for restoring data. That will just lead to terrible issues later. Please rethink this so we don't need to overly special-case things.

linked0 replied:
> No. As I said countless times, we shouldn't have a special path for restoring data. That will just lead to terrible issues later. Please rethink this so we don't need to overly special-case things.

I will rewrite the code in the morning as you commented. It would be good to have a pair-programming if you have time today.

linked0 replied:
> No. As I said countless times, we shouldn't have a special path for restoring data. That will just lead to terrible issues later. Please rethink this so we don't need to overly special-case things.

I have changed the approach in the last commit. Please tell me how it is. The commit for network test code is removed to avoid confusion and will be contained again soon.

Still need a test.

linked0 commented:
The commit for network test code is added again.

linked0 commented:
I'm seeing why the ci/circleci build is failing continuously.
Could you review my PR with the build failed?

Anyway, my code is totally reworked by following @Geod24's suggestions. The code has been simpler than the previous one.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -253,10 +253,12 @@ public class Ledger
         bool validators_changed = block.header.enrollments.length > 0
             || this.enroll_man.validatorCount() != old_count;
 
+        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
```
This should all be moved inside of `EnrollmentManager`. Because now we have enrollment logic living in both here and also in the EnrollmentManager.

Also what does the Ledger know about the node? It doesn't know whether it's a FullNode or a Validator, so this `getEnrollmentPublicKey` call is out of place here..

Here's one way to improve the code:

```diff
diff --git a/source/agora/consensus/EnrollmentManager.d b/source/agora/consensus/EnrollmentManager.d
index b60ac75d0..1619e824b 100644
--- a/source/agora/consensus/EnrollmentManager.d
+++ b/source/agora/consensus/EnrollmentManager.d
@@ -208,15 +208,17 @@ public class EnrollmentManager
     ***************************************************************************/
 
     public string addValidator (const ref Enrollment enroll,
-        Height block_height, scope UTXOFinder finder,
-        const ref UTXOSetValue[Hash] self_utxos) @safe nothrow
+        Height block_height, scope UTXOFinder finder) @safe nothrow
     {
         this.enroll_pool.remove(enroll.utxo_key);
 
         if (auto r = this.validator_set.add(block_height, finder, enroll))
             return r;
 
-        if (enroll.utxo_key in self_utxos)
+        const pubkey = this.getEnrollmentPublicKey();
+        UTXOSetValue utxo;
+        if (finder(enroll.utxo_key, size_t.max, utxo) &&
+            utxo.output.address == pubkey)
         {
             this.data = enroll;
             this.enroll_key = enroll.utxo_key;
@@ -479,7 +481,7 @@ public class EnrollmentManager
     ***************************************************************************/
 
     public void restoreValidators (Height last_height, const ref Block block,
-        scope UTXOFinder finder, const ref UTXOSetValue[Hash] self_utxos)
+        scope UTXOFinder finder)
         @safe nothrow
     {
         assert(last_height >= block.header.height);
@@ -487,8 +489,7 @@ public class EnrollmentManager
         {
             foreach (const ref enroll; block.header.enrollments)
             {
-                this.addValidator(enroll, block.header.height, finder,
-                    self_utxos);
+                this.addValidator(enroll, block.header.height, finder);
             }
         }
     }
@@ -791,9 +792,9 @@ unittest
 
     // test for enrollment block height update
     assert(man.getEnrolledHeight(utxo_hash) == ulong.max);
-    assert(man.addValidator(enroll, Height(9), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(9), &storage.findUTXO) is null);
     assert(man.getEnrolledHeight(enroll.utxo_key) == 9);
-    assert(man.addValidator(enroll, Height(9), &storage.findUTXO, utxos) !is null);
+    assert(man.addValidator(enroll, Height(9), &storage.findUTXO) !is null);
     assert(man.getEnrolledHeight(enroll2.utxo_key) == ulong.max);
     man.pool.getEnrollments(enrolls);
     assert(enrolls.length == 1);
@@ -826,7 +827,7 @@ unittest
     // So, a pre-image can only be got from the start height.
     PreImageInfo preimage;
     assert(man.createEnrollment(utxo_hash, Height(1), enroll));
-    assert(man.addValidator(enroll, Height(10), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(10), &storage.findUTXO) is null);
     assert(!man.getPreimage(Height(10), preimage));
     assert(man.getPreimage(Height(11), preimage));
     assert(preimage.hash == man.cycle.preimages[$ - 1]);
@@ -850,7 +851,7 @@ unittest
     // validator A with the `utxo_hash` and the enrolled height of 10.
     // validator B with the 'utxo_hash2' and the enrolled height of 11.
     // validator C with the 'utxo_hash3' and no enrolled height.
-    assert(man.addValidator(enroll2, Height(11), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll2, Height(11), &storage.findUTXO) is null);
     man.clearExpiredValidators(Height(11));
     assert(man.validatorCount() == 2);
     assert(man.getEnrolledUTXOs(keys));
@@ -859,7 +860,7 @@ unittest
     // set an enrolled height for validator C
     // set the block height to 1019, which means validator B is expired.
     // there is only one validator in the middle of 1020th block being made.
-    assert(man.addValidator(enroll3, Height(1019), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll3, Height(1019), &storage.findUTXO) is null);
     man.clearExpiredValidators(Height(1019));
     assert(man.validatorCount() == 1);
     assert(man.getEnrolledUTXOs(keys));
@@ -911,7 +912,7 @@ unittest
     assert(man.params.ValidatorCycle - 101 == 907); // Sanity check
     assert(man.getValidatorPreimage(utxo_hash) == PreImageInfo.init);
     auto preimage = PreImageInfo(utxo_hash, man.cycle.preimages[100], 907);
-    assert(man.addValidator(enroll, Height(2), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(2), &storage.findUTXO) is null);
     assert(man.addPreimage(preimage));
     assert(man.getValidatorPreimage(utxo_hash) == preimage);
 }
@@ -1033,7 +1034,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 1);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 3;
 
@@ -1044,7 +1045,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 2);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 4;
 
@@ -1055,7 +1056,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 3);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 5;    // valid block height : 0 <= H < 1008
     man.clearExpiredValidators(block_height);
@@ -1122,13 +1123,13 @@ unittest
     // Now create our own Enrollment
     // This means adding enrollment data to the block at the height of 2, 5, 7
     assert(man.createEnrollment(utxo_keys[0], Height(1), enroll));
-    man.addValidator(enroll, Height(2), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(2), &utxo_set.findUTXO);
     enrolls ~= enroll;
     assert(man.createEnrollment(utxo_keys[1], Height(2), enroll));
-    man.addValidator(enroll, Height(5), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(5), &utxo_set.findUTXO);
     enrolls ~= enroll;
     assert(man.createEnrollment(utxo_keys[2], Height(3), enroll));
-    man.addValidator(enroll, Height(7), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(7), &utxo_set.findUTXO);
     enrolls ~= enroll;
 
     PreImageInfo org_preimage;
@@ -1136,9 +1137,9 @@ unittest
 
     // clear up the EnrollmentManager for test
     man = new EnrollmentManager(":memory:", gen_key_pair, params);
-    man.addValidator(enrolls[0], Height(2), &utxo_set.findUTXO, utxos);
-    man.addValidator(enrolls[1], Height(5), &utxo_set.findUTXO, utxos);
-    man.addValidator(enrolls[2], Height(7), &utxo_set.findUTXO, utxos);
+    man.addValidator(enrolls[0], Height(2), &utxo_set.findUTXO);
+    man.addValidator(enrolls[1], Height(5), &utxo_set.findUTXO);
+    man.addValidator(enrolls[2], Height(7), &utxo_set.findUTXO);
 
     // check the restored enrollment data
     assert(man.enroll_key == enrolls[2].utxo_key);
diff --git a/source/agora/node/Ledger.d b/source/agora/node/Ledger.d
index 0f55bd7ee..78a140f0c 100644
--- a/source/agora/node/Ledger.d
+++ b/source/agora/node/Ledger.d
@@ -137,14 +137,12 @@ public class Ledger
 
         // restore validator set from the blockchain.
         // using block_count, as the range is inclusive
-        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
         foreach (block_idx; min_height .. block_count)
         {
             Block block;
             this.storage.readBlock(block, block_idx);
-            UTXOSetValue[Hash] utxos = this.utxo_set.getUTXOs(pubkey);
             this.enroll_man.restoreValidators(this.last_block.header.height,
-                block, this.utxo_set.getUTXOFinder(), utxos);
+                block, this.utxo_set.getUTXOFinder());
         }
     }
 
@@ -255,12 +253,10 @@ public class Ledger
         bool validators_changed = block.header.enrollments.length > 0
             || this.enroll_man.validatorCount() != old_count;
 
-        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
-        UTXOSetValue[Hash] utxos = this.utxo_set.getUTXOs(pubkey);
         foreach (idx, ref enrollment; block.header.enrollments)
         {
             if (auto r = this.enroll_man.addValidator(enrollment,
-                block.header.height, this.utxo_set.getUTXOFinder(), utxos))
+                block.header.height, this.utxo_set.getUTXOFinder()))
             {
                 log.fatal("Error while adding a new validator: {}", r);
                 log.fatal("Enrollment #{}: {}", idx, enrollment);
```

linked0 replied:
> Here's one way to improve the code:
```
+        if (finder(enroll.utxo_key, size_t.max, utxo) &&
+            utxo.output.address == pubkey)
```
The suggestion by @AndrejMitrovic didn't work for my revised network test which is in the last commit. So, I have returned to my original code about that.

AndrejMitrovic replied:
Why? What was failing?

linked0 replied:
> Why? What was failing?

The code for adding validators has failed with the `finder`.  You can check that with my last commit having the network test. I guess it's because the `finder` function put a UTXOSetValue passed through a parameter into the `used_utxos'.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,52 @@ unittest
+    network.waitForDiscovery();
+    containSameBlocks(nodes, 0).retryFor(5.seconds);
+
+    // Check if the validator information had been stored from the second block.
```
Which second block? There was no second block generated in this test.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,52 @@ unittest
+    auto enroll = b0.header.enrollments[0];
+
+    // The `enroll` hasn't been validator yet so there must not be a pre-image
+    assert(node_1.getPreimage(enroll.utxo_key) != PreImageInfo.init);
```
The comment and the code differ here. The comment says "there should not be a pre-image", but here you are testing for `!= PreImageInfo`, which means there is a preimage.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,52 @@ unittest
+    PreImageInfo preimage_1 = node_1.getPreimage(enroll.utxo_key);
+    assert(preimage_1 == org_preimage);
+
+    // Wait for the timer for revelation of a pre-image to elaspe
```
```suggestion
    // Wait for the timer for revelation of a pre-image to elapse
```

linked0 commented:
@AndrejMitrovic Thank you for your review. I'll apply the suggestions to it.

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -689,22 +700,25 @@ unittest
     scope storage = new TestUTXOSet;
 
     auto gen_key_pair = WK.Keys.Genesis;
+    UTXOSetValue[Hash] utxos;
```
These changes should also be reverted because the utxo set here is not used anymore.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1045,3 +1068,75 @@ unittest
+    import std.conv;
+
+    scope utxo_set = new TestUTXOSet;
+    UTXOSetValue[Hash] utxos;
```
This `utxos` should also be removed.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -475,13 +475,22 @@ public class EnrollmentManager
             last_height = the latest block height
             block = the block to update the validator set with
             finder = the delegate to find UTXOs with
+            self_utxos = the UTXOs belonging to this node
```
Unused.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -137,12 +137,14 @@ public class Ledger
         {
             Block block;
             this.storage.readBlock(block, block_idx);
+            UTXOSetValue[Hash] utxos = this.utxo_set.getUTXOs(pubkey);
```
Unused.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -137,12 +137,14 @@ public class Ledger
 
         // restore validator set from the blockchain.
         // using block_count, as the range is inclusive
+        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
```
Unused.

AndrejMitrovic commented on source/agora/consensus/PreImage.d:
```diff
@@ -343,15 +343,18 @@ public struct PreImageCycle
-                   to generate the cycle seeds
+            secret = The secret key of the node, used as part of the hash
+                to generate the cycle seeds
+            calc_index = The flag to determine to calculating the cycle index
+                or just getting a random seed, or the first pre-image
```
I'm a little bit confused by this sentence because it's saying it will do one of **three** things but it's a boolean.

linked0 replied:
> I'm a little bit confused by this sentence because it's saying it will do one of **three** things but it's a boolean.

The description should be changed. Thanks.

linked0 commented:
My code fails with this output sometimes.
```
(base) [~/work/agora]$ lldb ./build/agora-unittests                                                                                         [remove-enroll-table]
(lldb) target create "./build/agora-unittests"
Current executable set to '/Users/jaylee/work/agora/build/agora-unittests' (x86_64).
(lldb) r
Process 4744 launched: '/Users/jaylee/work/agora/build/agora-unittests' (x86_64)
Process 4744 stopped
* thread #12, stop reason = EXC_BAD_ACCESS (code=1, address=0x63e9df8f30)
    frame #0: 0x0000000100169963 agora-unittests`_D5agora7network13NetworkClientQp12sendPreimageMFNeKSQBz9consensus4data12PreImageInfoQoZ9__lambda2MFZ12__dgliteral1MFNaNiNfZv at NetworkClient.d:325:57
   322 	    {
   323 	        this.taskman.runTask(
   324 	        {
-> 325 	            this.attemptRequest(this.api.receivePreimage(preimage), null);
   326 	        });
   327 	    }
   328 	}
Target 0: (agora-unittests) stopped.
```
If you have some clues for that, please let me know. Anyway, I'm now finding the cause for that.

AndrejMitrovic commented:
That bug has been there for a while, and it triggers randomly with some PRs. It's often in `sendPreimage`:

```
* thread #2, stop reason = EXC_BAD_ACCESS (code=1, address=0x40fc8c4f50)
  * frame #0: 0x0000000100168b93 agora-unittests`_D5agora7network13NetworkClientQp12sendPreimageMFNeKSQBz9consensus4data12PreImageInfoQoZ9__lambda2MFZ12__dgliteral1MFNaNiNfZv at random.d:2303:9
    frame #1: 0x000000010016810c agora-unittests`_D5agora7network13NetworkClientQp__T14attemptRequestVE5ocean4util3log7ILoggerQi5Leveli0TvZQCcMFLvC9ExceptionZv(this=0x000000010b946e40, dg=void delegate() @ 0x00000041098d2f68, ex=0x0000000000000000) at NetworkClient.d:259:15
    frame #2: 0x0000000100168b4d agora-unittests`_D5agora7network13NetworkClientQp12sendPreimageMFNeKSQBz9consensus4data12PreImageInfoQoZ9__lambda2MFZv at random.d:0:17
    frame #3: 0x000000010069b579 agora-unittests`_D6geod2411concurrency14FiberScheduler6createMFNbDFZvZ4wrapMFZv at concurrency.d:916:13
    frame #4: 0x0000000100af6a89 agora-unittests`fiber_entryPoint + 89
```

It was taking me too much time to find the root cause, so I haven't found what's causing it.

linked0 commented:
The segmentation fault is resolved in the PR #975 .

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -193,28 +193,42 @@ public class EnrollmentManager
         Add a validator to the validator set or update the enrolled height.
 
+        This also gets and stores the information about this node's enrollment
+        information which could be lost in abnormal situations.
```
Side note: What do you consider "abnormal situation" ? For me a node restarting is not abnormal. Nor is a node taking over another one.
What could be considered abnormal would be a catastrophic failure such as "disk unplugged". But "abnormal" makes me think it things that shouldn't happen, when they can and will happen, and we need to cater for it. Even if we equate "abnormal" to "catastrophic events", I think it is much better to describe what event, or what effect they would have (e.g. for "disk unplugged": inability to read / write to disk) than just calling it "catastrophic event".

But it's just a side note for future reference. I don't think it's worth fixing right now.

Geod24 commented on source/agora/consensus/PreImage.d:
```diff
@@ -343,15 +343,18 @@ public struct PreImageCycle
     ***************************************************************************/
 
-    public Hash consume (scope const ref Scalar secret) @safe nothrow
+    public Hash consume (scope const ref Scalar secret, bool calc_index = true)
+        @safe nothrow
```
Since the name is `consume`, the `calc_index` argument is at odds with the goal of the function.
I would rename it to `populate` and rename `calc_index` to `consume`.
The description also needs to be updated.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -687,24 +699,26 @@ unittest
+        auto tx_hash = hashFull(tx);
+        auto signature = gen_key_pair.secret.sign(tx_hash[]);
         tx.inputs[0].signature = signature;
         storage.put(tx);
+        auto ut_hash = UTXOSetValue.getHash(tx_hash, 0);
+        utxos[ut_hash] = UTXOSetValue(1, TxType.Freeze, output);
```
This creates an invalid state. The `UTXOSetValue.unlock_height` shouldn't be `1`.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -853,23 +867,27 @@ unittest
         tx.inputs[0].signature = signature;
         storage.put(tx);
+        auto ut_hash = UTXOSetValue.getHash(tx_hash, 0);
+        utxos[ut_hash] = UTXOSetValue(1, TxType.Freeze, output);
```
Another invalid state

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -966,24 +985,28 @@ unittest
         storage.put(tx);
+
+        auto utxo_hash = UTXOSetValue.getHash(tx_hash, 0);
+        utxos[utxo_hash] = UTXOSetValue(1, TxType.Freeze, output);
```
Invalid state

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1045,3 +1068,74 @@ unittest
+    auto key_pair = KeyPair.random();
+
+    // TODO: makeChainedTransactions does not yet support frozen txs
+    foreach (idx; 0 .. 7)
+    {
+        auto input = Input(hashFull(GenesisTransaction), idx.to!uint);
+        auto output = Output(Amount.MinFreezeAmount, gen_key_pair.address);
+        Transaction tx =
+        {
+            TxType.Freeze,
+            [input],
+            [output]
+        };
+
+        auto tx_hash = hashFull(tx);
+        auto signature = gen_key_pair.secret.sign(tx_hash[]);
+        tx.inputs[0].signature = signature;
+        utxo_set.put(tx);
```
```suggestion
    foreach (uint idx; 0 .. 7)
    {
        auto tx = TxBuilder(GenesisTransaction, idx).draw(Amount.MinFreezeAmount, key_pair).sign(TxType.Freeze));
        utxo_set.put(tx);
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1045,3 +1068,74 @@ unittest
+        utxo_set.put(tx);
+
+        auto utxo_hash = UTXOSetValue.getHash(tx_hash, 0);
+        utxos[utxo_hash] = UTXOSetValue(1, TxType.Freeze, output);
```
Invalid state

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1045,3 +1068,74 @@ unittest
+    auto man = new EnrollmentManager(":memory:", gen_key_pair, params);
+
+    // create and add enrollments to the ValidatorSet at the height of 2, 5, 7
+    Enrollment enroll;
+    Enrollment[] enrolls;
+    Hash[] utxo_keys = utxo_set.keys;
+    assert(man.createEnrollment(utxo_keys[0], Height(1), enroll));
+    man.addValidator(enroll, Height(2), &utxo_set.findUTXO, utxos);
+    enrolls ~= enroll;
+    assert(man.createEnrollment(utxo_keys[1], Height(2), enroll));
+    man.addValidator(enroll, Height(5), &utxo_set.findUTXO, utxos);
+    enrolls ~= enroll;
+    assert(man.createEnrollment(utxo_keys[2], Height(3), enroll));
+    man.addValidator(enroll, Height(7), &utxo_set.findUTXO, utxos);
+    enrolls ~= enroll;
+
+    PreImageInfo org_preimage;
+    assert(man.getPreimage(Height(10), org_preimage));
+
+    // clear up the EnrollmentManager for test
+    man = new EnrollmentManager(":memory:", gen_key_pair, params);
+    man.addValidator(enrolls[0], Height(2), &utxo_set.findUTXO, utxos);
+    man.addValidator(enrolls[1], Height(5), &utxo_set.findUTXO, utxos);
+    man.addValidator(enrolls[2], Height(7), &utxo_set.findUTXO, utxos);
+
+    // check the restored enrollment data
+    assert(man.enroll_key == enrolls[2].utxo_key);
+    assert(man.data == enrolls[2]);
+
+    // check the restored pre-image
+    PreImageInfo preimage;
+    assert(man.getPreimage(Height(10), preimage));
+    assert(preimage == org_preimage);
```
The return of `addValidator` isn't checked through the test.

Geod24 commented on source/agora/consensus/PreImage.d:
```diff
@@ -343,15 +343,18 @@ public struct PreImageCycle
         Params:
-          secret = The secret key of the node, used as part of the hash
-                   to generate the cycle seeds
+            secret = The secret key of the node, used as part of the hash
+                to generate the cycle seeds
```
Undo

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,50 @@ unittest
+
+    // Check if genesis block has enrollments
+    const b0 = node_1.getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length == 4);
```
```suggestion
    assert(b0.header.enrollments.length >= 1);
```

Since you are only using one, let's simplify this. This also ensures that if we extend the genesis block, this assert doesn't trigger needlessly.

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,50 @@ unittest
+    assert(b0.header.enrollments.length == 4);
+    auto enroll = b0.header.enrollments[0];
+
+    // Check if the `node_1` restores block from the chain after restarting
+    network.restart(node_1);
+    network.waitForDiscovery();
+    containSameBlocks(nodes, 0).retryFor(5.seconds);
```
That's not really checking anything. We haven't created any block, so there is nothing to restore.

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -96,3 +96,50 @@ unittest
+    PreImageInfo preimage_1 = node_1.getPreimage(enroll.utxo_key);
+    assert(preimage_1 == org_preimage);
+
+    // Wait for the timer for revelation of a pre-image to elapse
+    Thread.sleep(10.seconds);
```
We have `retryFor` for this.

Geod24 reviewed (changes requested):
The code is very nice, and so are the commit message, thanks a lot for the work you put in this!
The test, however, does not seem to be testing anything: Since the enrollments are in the genesis block, they will always get restored no matter what (because the block is available when the constructor is run).

linked0 commented:
Thank you for your review.
I will add the test code for abnormal situations. My first focus in the network test was to restore a node's enrollment information from blocks and reveal the next pre-image. So I have a plan to write another network test code about an abnormal situation we discussed in another PR in this sprint.

linked0 commented:
It keeps failing in ci/circleci from yesterday evening. 
Could you review this PR though it fails? or Please let me know if there is what I can do.

AndrejMitrovic commented:
There seem to be a lot of duplicate requests? E.g.:

```
node-0_1_be5208273588 | 2020-06-24T00:38:04.068570582Z 2020-06-24 00:38:04,58 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.200557928Z 2020-06-24 00:38:04,68 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.200593099Z 2020-06-24 00:38:04,103 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.200600568Z 2020-06-24 00:38:04,179 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.230039202Z 2020-06-24 00:38:04,200 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.230073761Z 2020-06-24 00:38:04,225 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.230242372Z 2020-06-24 00:38:04,228 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.350240053Z 2020-06-24 00:38:04,229 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.350275439Z 2020-06-24 00:38:04,293 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.350297738Z 2020-06-24 00:38:04,347 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
node-0_1_be5208273588 | 2020-06-24T00:38:04.381063595Z 2020-06-24 00:38:04,350 Trace [agora.node.FullNode] - Received Preimage: { enroll_key: 0xf21f606e96d6130b02a807655fda22c8888111f2045c0d45eda9c26d3c97741ca32fc68960ae68220809843d92671083e32395a848203380e5dfd46e4b0261f0, hash: 0xe94904bf1e2aa5444df4bc39992f6b97d98a071a8e2268cc74807256f0a599b8c142f55e94760acc38f02d5c691e73d8ed67be5d7d3c74935ee30f14e4f5da49, distance: 11 }
```

It seems one preimage is constantly sent out, very frequently.

Geod24 commented:
Concept LGTM, let's fix the details together.

linked0 commented:
> It seems one preimage is constantly sent out, very frequently.

I think I can find where the issues are. If needed I will create a new issue. Thank you for the information.
Anyway, how can I find the log as you showed in the local environment?

AndrejMitrovic commented:
> Anyway, how can I find the log as you showed in the local environment?

It was on the CircleCI page.

Geod24 commented:
> The test, however, does not seem to be testing anything: Since the enrollments are in the genesis block, they will always get restored no matter what (because the block is available when the constructor is run).

This is still valid.

linked0 commented:
> > The test, however, does not seem to be testing anything: Since the enrollments are in the genesis block, they will always get restored no matter what (because the block is available when the constructor is run).
> 
> This is still valid.

Sorry about forgetting that.
I will add more code about that in the third commit, "Network test for restoring a node's enrollment".

linked0 commented:
I have a log as follows about the Validator #1 node.
Could it mean that the node crashed after the 20th block had been added in the chain? I can see the message "Doing network discovery...".
```
2020-06-26 08:27:07,433 [32mInfo[0m [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Proposing tx set for slot 20
2020-06-26 08:27:07,435 [32mInfo[0m [agora.consensus.protocol.Nominator] - agora.consensus.protocol.Nominator.Nominator.nominate(): Tx set nominated
2020-06-26 08:27:07,447 [32mInfo[0m [agora.consensus.protocol.Nominator] - combineCandidates: cast(__c_ulonglong) 20
2020-06-26 08:27:07,537 [32mInfo[0m [agora.consensus.protocol.Nominator] - Externalized consensus data set at cast(__c_ulonglong) 20: { tx_set: { _set: [ { type: TxType.Payment, inputs: [{ previous: 0x54ab80f9225af2c2977485cbcaea869f0dcb282991985f341f60a3955a2fbb9db3f5d3e656232254b299acefdf831c39726242c12976ed95635fc2955d2bd478, index: 0, signature: 0x0e94e6c0cd9e02fccd0c16687fbf7a6d3ef7159efe15191c4fc45a16d77f7df345e90f86236c4d5a1935e87455b8e8cfe257383727719615d54ff05ade85fff1 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0x9361454564d8c5d68dd632d7b8cdf7bf827e97e85ff7f7066eac0f065b0f3c8f9364d2763ba178ba130dfb7a75e7fca0c28575cbd47df7f6d43465280a0798bf, index: 0, signature: 0x05a49aef7cdd71e5360324482f4b35e9ce50ac8b39d24bbb8684cc7cae532d67977c7a2c9c5ce9b66940971d8ad1377405d630bb481a21e18dcab87fd9f73c86 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0xd8bb2d79967213a488aa8c67d5fbc61556975c906722814204046cc3f155c88146092588eb4e06c5c8bf1c05ae98c1448f862022bbc792930a4f67d4016d660a, index: 0, signature: 0x050e6235ca6c3fe532f0f4e71fed6a753f51d2e86b97cc38dd7ff8078a14d48458b5a48433cd9ec73d5ac782c7ac800f4d66064285a1e2643a5a1b1f245c5ea0 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0xdd3a6049e6ecdd2027c1f1327ab715ed178f9302882d17fa7d62b13d69c673fe9dc7ba542967ed7af35161e568c34d71954b7833fcd784fece5c2f183172c796, index: 0, signature: 0x0a7fbe50aa1063462bd04ca7bd3aa1ec9ff99b282b9982dfff8e2169ac121a942eb93d73b1e2de24cdf9de82b9854bdfce43ceb78f63cc60f0d16111a668c1fc }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0xc3262b4c76ed148cca22ae7ee006b68a5d2c83e9693caf98d56b4638bd8aa1f98a2ba50b16611f377fb2ee489d5e629dd12aa013ed43887f103e8999718bcc83, index: 0, signature: 0x0b69087212f3d911ec22c22f2f3fcd69f3be281ea4cfb5cd5e3f72b9240096784588ba3b792a1277e10e6c248d5aa04db550c7d62a5d45666e20214a8f5a8fd6 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0xd7b97a758b88d52a3187d4d3b35ac5168a5cd7907ea86a36605c58eed87cbd04eadcf9d57a8faf27d6310023728e6d64cda9d535b0e084b0721ed5b1988d135f, index: 0, signature: 0x0486f1c4eaeb612b2193250e01f571f0e9afeb950bf6a557e543ae1d7e8631ef764630eaaf6edd18c399659e39418f6c7eed117a0056867cb7cefb6dbe2d7dd0 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0x6afc25c4ec800730192cd635840e7f2ab4b014842529275b929b82a9f2989df7a766577be22d6721f8de90c0f6631522ba37fe1834c9f821aba11fa279a66db2, index: 0, signature: 0x0088377a978f95293f9b6a92caa69937c5b5e93687b94ed1203ac24a75c1d35090a7e2c49ec896f91aba3ddf71a0655face26d37dd9999a6a095a7eb169e2605 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true, { type: TxType.Payment, inputs: [{ previous: 0x31a1c281f7512f2e3fe647ef94e0f8551c3214906fb57bb16d2bc28671d37f991f825b1765fc79ff40d154937e5fbe878707a0ab785a49839cbe6dddd7e5af6e, index: 0, signature: 0x0e7fa6d839e6945f81b07a1113124ef0b00be223861b8483bb564ba1e0776412499de816a91aba67c47c68730db49802e6710b3c567c56d4ebeb0f4963dae2b6 }], outputs: [{ value: 625000000000000, address: GDU22IOIYY2RDYILBCWAFNSGV2GFTS5NZXWY2LV77UCOJUSQCAGYCMZI }] }: true ] }, enrolls: [{ utxo_key: 0xb82cb96710af2e9804c59d1f1e1679f8b8b69f4c0f6cd79c8c12f365dd766c09aaa4febcc18b3665d33301cb248ac7afd343ac7b98b27beaf246ad12d3b3219a, random_seed: 0xd87ffdf613073bd0309a9002559dba0ad5a3a7b4ad545af85aafb077810a8382f29550a18620570cc5a7808cac965f853de204c2a054ceeb918c5d01283db028, cycle_length: 20, enroll_sig: 0x0524a98e0a0ef0c4b7eb3fab8879d1c5cf57af4f8f2bc9d05ab74a35a6d8b03b7193a6bf900ab9eee54b5ea0917d8881bfeed52317dfba0ac07fc63b07cc84c7 }] }
2020-06-26 08:27:07,539 [36mTrace[0m [agora.consensus.protocol.Nominator] - Rejected envelope with outdated slot #cast(const(__c_ulonglong)) 20 (ledger: #{ value: 20 })
2020-06-26 08:27:07,539 [36mTrace[0m [agora.consensus.protocol.Nominator] - Rejected envelope with outdated slot #cast(const(__c_ulonglong)) 20 (ledger: #{ value: 20 })
2020-06-26 08:27:07,540 [36mTrace[0m [agora.consensus.protocol.Nominator] - Rejected envelope with outdated slot #cast(const(__c_ulonglong)) 20 (ledger: #{ value: 20 })
2020-06-26 08:27:10,400 [32mInfo[0m [agora.network.NetworkManager] - Doing network discovery..n: { value: { type: const(TxType).Payment, inputs: [{ previous: 0xdb8e0b39dc0ca8471e7105d40faa049fe7ac24d0e2deacb02d07e2448122cc09b8d54d179726b3d9051dd2e1f17255e0ca400c60469f12b6a9e753248ab78e17, index: 0, signature: 0x09d570a2dd162a9692ba4dd0fd7e3135361842763d27d88883c42108334031e019b1513c24fdbc58faa863b966be637ec043c248a3dcf7b0cb92620279eb14ad }], outputs: [{ value: 625000000000000, address: GDN22BSZ6JCLELE4AJJJR4DYSIGK72Q37RLAEX2AH7CFULG4OUQB6A7I }] } }
```

linked0 commented:
The error in some CI builds and the error that occurs sometimes in my local machine is related?

Some CI builds fail with a log as follows.
```
Linking...
Running ./build/agora-unittests 
0   agora-unittests                     0x0000000104ffa15e _D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv + 62
Program exited with code -11
##[error]Process completed with exit code 2.
```

On my machine, `dub test` fails sometimes with a log as follows.
```
Running ./build/agora-unittests
#### INTERNAL ERROR: Range violation
This node was started at source/agora/test/Base.d:406
Please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.RangeError@submodules/localrest/source/geod24/LocalRest.d(591): Range violation
----------------
??:? onRangeError [0x10c2dbdf8]
??:? _d_arraybounds [0x10c2dc647]
??:? void geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI, agora.test.Base.Serializer).RemoteAPI.spawned!(agora.test.Base.TestValidatorNode).spawned(geod24.concurrency.Tid, immutable(char)[], int, agora.common.Config.Config, shared(geod24.Registry.Registry)*, immutable(agora.consensus.data.Block.Block)[], immutable(agora.consensus.data.ConsensusParams.ConsensusParams)).runNode().handle!(geod24.LocalRest.Response).handle(geod24.LocalRest.Response) [0x10b8fe3aa]

...

Module tests failed: agora.test.EnrollmentManager
core.exception.AssertError@source/agora/test/EnrollmentManager.d(194): Check condition failed after timeout of 5 secs and 50 attempts: Node Validator #0 (GDA2...FKWI) has not completed discovery after 5 secs.
----------------
??:? void agora.utils.Test.retryFor!(core.exception.AssertError).retryFor(lazy bool, core.time.Duration, lazy immutable(char)[], immutable(char)[], ulong) [0x10b89ca1e]
??:? void agora.test.Base.TestAPIManager.waitForDiscovery(immutable(char)[], ulong).__lambda3!(agora.test.Base.NodePair).__lambda3(agora.test.Base.NodePair) [0x10b901b62]
??:? std.typecons.Flag!("each").Flag std.algorithm.iteration.each!(agora.test.Base.TestAPIManager.waitForDiscovery(immutable(char)[], ulong).__lambda3).each!(agora.test.Base.NodePair[]).each(ref agora.test.Base.NodePair[]) [0x10b8e6c65]
??:? void agora.test.Base.TestAPIManager.waitForDiscovery(immutable(char)[], ulong) [0x10b8ca9e6]
??:? void agora.test.EnrollmentManager.__unittest_L148_C1() [0x10b8e09b0]
??:? agora.test.EnrollmentManager.__unittest [0x10b8e458f]
??:? void agora.test.Base.customModuleUnitTester().runTest(agora.test.Base.customModuleUnitTester().ModTest) [0x10b8e4c7a]
??:? int agora.test.Base.customModuleUnitTester().__foreachbody4(ref agora.test.Base.customModuleUnitTester().ModTest) [0x10b8e5ae0]
??:? void std.parallelism.ParallelForeach!(agora.test.Base.customModuleUnitTester().ModTest[]).ParallelForeach.opApply(scope int delegate(ref agora.test.Base.customModuleUnitTester().ModTest)).doIt() [0x10bd071c1]
??:? void std.parallelism.TaskPool.doJob(std.parallelism.AbstractTask*) [0x10c25fef0]
??:? void std.parallelism.TaskPool.startWorkLoop() [0x10c26009a]
??:? thread_entryPoint [0x10c2e5654]
??:? _pthread_start [0x7fff6c095108]
```

linked0 commented:
Sometimes the error as follows occurs on some CI build.
```
Linking...
Running ./build/agora-unittests 
0   agora-unittests                     0x000000011083067e _D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv + 62
1   libsystem_platform.dylib            0x00007fff6b2d95fd _sigtramp + 29
Program exited with code -10
##[error]Process completed with exit code 2.
```

linked0 commented:
Ready for review.
I'm concerned about the above [failure](https://github.com/bpfkorea/agora/pull/798#issuecomment-654649711) on some CI build.
After this PR, I will add the column for the cycle index of the random seeds into the `node_enroll_data`.

AndrejMitrovic commented:
> Relates #789

Does this fix #789 or it's just the first part? If it fixes it then "Fixes #789" should be in one of the commit messages.

linked0 commented:
> Does this fix #789 or it's just the first part? If it fixes it then "Fixes #789" should be in one of the commit messages.

I will make another PR today which could have the "Fixes #789". Thanks.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -228,28 +228,41 @@ public class EnrollmentManager
 
```
The commit message should mention that `consume` was renamed to `populate`

AndrejMitrovic commented on source/agora/consensus/PreImage.d:
```diff
@@ -332,26 +332,30 @@ public struct PreImageCycle
           secret = The secret key of the node, used as part of the hash
                    to generate the cycle seeds
+            consume = The flag to determine to calculating the cycle index
+                or just getting a random seed which is the first pre-image
```
The documentation should be explicit about which value will do what action, for example:

```suggestion
          consume = if true, calculate the cycle index. Otherwise, just get the random seed which is the first pre-image
```

linked0 replied:
> The documentation should be explicit about which value will do what action, for example:

I have changed it.

AndrejMitrovic commented on source/agora/consensus/PreImage.d:
```diff
@@ -332,26 +332,30 @@ public struct PreImageCycle
-        is part of consensus.
+        so `populate` needs to be called once this node has confirmed it
+        is part of consensus. If the caches are only needed without the
+        increment, the `consume` parameter must be `false`.
```
```suggestion
        increment, the `consume` parameter must be set to `false`.
```

linked0 replied:
> increment, the `consume` parameter must be set to `false`.

I have changed it.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -277,7 +290,7 @@ public class EnrollmentManager
         // TODO: Move this consume call to `addValidator` and instead
```
This comment is now outdated.

linked0 replied:
> This comment is now outdated.
I removed it.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -449,7 +449,11 @@ public class EnrollmentManager
             this.validator_set.getEnrolledHeight(this.enroll_key);
         // FIXME: This function should only be called if we're enrolled
-        assert(enrolled_height != ulong.max);
+        if (enrolled_height == ulong.max)
```
That's a good change, but what about a unittest?

linked0 replied:
> That's a good change, but what about a unittest?

I added the test.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -457,9 +457,10 @@ public class EnrollmentManager
         const start_height = enrolled_height + 1;
```
> Sometimes, the requested index of pre-image is above the cycle minus 1

Can you add a small note when this would actually happen?

Also there is a typo: munus => minus

linked0 replied:
> Can you add a small note when this would actually happen?
> Also there is a typo: munus => minus

I added the note to commit message and fix the typo.

linked0 replied:
> Can you add a small note when this would actually happen?
> Also there is a typo: munus => minus

I added the note to commit message and fix the typo.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +142,82 @@ unittest
+///     node has been inserted into a new block with consensus. And the node
+///     starts again
+/// Expectation: The node restores its own enrollment from the chain and
+///     reveals its pre-images preidically.
```
```suggestion
///     reveals its pre-images periodically.
```

linked0 replied:
> ///     reveals its pre-images preidically.

I have changed it.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +142,82 @@ unittest
+        .array();
+    txs.each!(tx => nodes[1].putTransaction(tx));
+    ensureConsistency(nodes.take(1), 20);
+    assert(nodes[0].getBlocksFrom(20, 1)[0].header.enrollments.length == 1);
```
I think this is better with `retryFor` in case the request times out (it means there will be less random test-suite failures).

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +142,82 @@ unittest
+    network.waitForDiscovery();
+    ensureConsistency([nodes[0]], 20, 10.seconds);
+
+    // Now make a new block and make sure only nodes[1] signs it
```
```suggestion
    // Now make a new block and make sure only nodes[0] signs it
```

Right?

linked0 replied:
> // Now make a new block and make sure only nodes[0] signs it
>
> Right?

The nodes[1] is right. We used that intentionally in the pair-programming with @Geod24

AndrejMitrovic replied:
Is the comment referring to signing the transaction or the block? That's what I was confused by the wording. "make a new block and sign it" refers to the block.

linked0 replied:
By the way, If I change the comment, does the code show the situation that only nodes[0] sign it?

AndrejMitrovic replied:
Sign what? That was my question.

linked0 replied:
It was written by @Geod24 so I think he intended block.

AndrejMitrovic replied:
Hmm.. But I think all of the node's enrollments have expired at block height 20, except node[0] which re-enrolled. So I'm not sure that comment was correct? Let's see what he says.

AndrejMitrovic replied:
Anyway there is no actual verification here of **which node** created the block. So "make sure" doesn't make sense here.

linked0 replied:
OK, I will modify it to the right expression after checking.

linked0 replied:
At last, I agree with you. It should be `nodes[0]`. Sorry for my late understanding. Thanks.

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +142,82 @@ unittest
+        10.seconds);
+
+    // Check if a new pre-image has been revealed from the restarted node
+    auto temp_hash = preimage_2.hash;
```
Could just use the preimage validation routine here?

AndrejMitrovic reviewed (changes requested):
Small notes.

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1096,4 +1161,7 @@ unittest
     assert(preimage == org_preimage);
+
+    // check the cycle count for enrollment
+    assert(man.loadIntegerData("cycle_count") == 3);
```
More test code or network test is needed.

linked0 commented:
I will request a review at green. I intended to write a new test code for saving and loading the cycle index for `PreImageCycle` but it is just related to database operation and the following code has a logical test, I guess. 

```
assert(man.addValidator(enrolls[0], Height(2), finder, utxos) is null);
assert(man.addValidator(enrolls[1], Height(5), finder, utxos) is null);
assert(man.addValidator(enrolls[2], Height(7), finder, utxos) is null);
...
// check the cycle count for enrollment
assert(man.loadIntegerData("cycle_count") == 3);
```
If you, guys request another test for that, I will add one ASAP. Thanks.

linked0 commented:
Ready for review.
I added a commit about saving the cycle index of the `PreImageCycle`.

The CI builds fail only on `ldc-mater` compiler of each OSs as follows.
```
Assertion failed: (global->isDeclaration() && "Global variable already defined"), function defineGlobal, file /Users/runner/work/1/s/gen/llvmhelpers.cpp, line 1806.
0  ldc2                     0x0000000103ac7a85 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  ldc2                     0x0000000103ac6836 llvm::sys::RunSignalHandlers() + 198
2  ldc2                     0x0000000103ac809c SignalHandler(int) + 268
3  libsystem_platform.dylib 0x00007fff6c4385fd _sigtramp + 29
4  libsystem_platform.dylib 0x00000001080a5080 _sigtramp + 18446603343129660064
5  libsystem_c.dylib        0x00007fff6c30a808 abort + 120
6  libsystem_c.dylib        0x00007fff6c309ac6 err + 0
7  ldc2                     0x0000000103dd23bc (anonymous namespace)::DeclareOrDefineVisitor::visit(TypeInfoStructDeclaration*) + 4636
8  ldc2                     0x0000000103d38654 CodegenVisitor::visit(TypeInfoDeclaration*) + 116
9  ldc2                     0x0000000103d698bf codegenModule(IRState*, Module*) + 4255
10 ldc2                     0x0000000103deb831 ldc::CodeGenerator::emit(Module*) + 881
11 ldc2                     0x0000000103e050b6 codegenModules(Array<Module*>&) + 822
12 ldc2                     0x0000000103cef7ed mars_mainBody(Param&, Array<char const*>&, Array<char const*>&) + 19245
ldc2 failed with exit code -6.
##[error]Process completed with exit code 2.
```

Geod24 commented:
Congrats, you found a compiler bug!

AndrejMitrovic commented:
Could you rebase this PR? It will get rid of LDC compilation errors.

AndrejMitrovic commented:
The `ensureConsistency` commit is in the wrong position. It is introduced after it is used in the commit `Network test for restoring a node's enrollment`.

AndrejMitrovic commented:
This might fix it:

```
git rebase -i origin/v0.x.x -x "git commit --amend --reset-author -CHEAD"
```

But only if it's in the right position locally. Otherwise you can reorder it with `git rebase -i` as usual.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -463,9 +463,10 @@ public class EnrollmentManager
-            return false;
+
+        // If the index is over the validator cycle, set the max value.
+        immutable index = (height - start_height <= this.params.ValidatorCycle - 1) ?
```
This could be:

```suggestion
        immutable index = min(height - start_height, this.params.ValidatorCycle - 1));
```

linked0 commented:
> ```
> git rebase -i origin/v0.x.x -x "git commit --amend --reset-author -CHEAD"
> ```
> 
> But only if it's in the right position locally. Otherwise you can reorder it with `git rebase -i` as usual.

This command is what I’m looking for these days, which was what @Geod24 introduced in some pair-programming session. Thanks.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -164,6 +164,12 @@ public class EnrollmentManager
+
+        // load the count of 'populate' of `PreImageCycle`
+        const uint populate_count =
+            cast(uint)this.loadIntegerData("cycle_count");
```
Why not change `loadIntegerData` to return a `uint`?

linked0 replied:
I intended that the function would be used for `getNextRevealHeight`.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -692,6 +702,61 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    private size_t loadIntegerData (string name, size_t default_value = 0)
```
`default_value` is unused by any call, why does it need to be a parameter?

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -692,6 +702,61 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    private size_t loadIntegerData (string name, size_t default_value = 0)
```
If these two methods are supposed to be generic (load/save "any" integer data), then why not also use them in `getNextRevealHeight` and `setNextRevealHeight`?

It seems inconsistent to have generic methods but they're only used in one call.

linked0 replied:
I will use it for the two functions in another commit soon.

linked0 commented:
I added new commit named "Use `load/saveIntegerData` for `next_reveal_height`" for @AndrejMitrovic 's suggestion.

Geod24 commented:
Needs a rebase now that your other PR has been merged.

linked0 commented:
Rebased.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -659,53 +693,77 @@ public class EnrollmentManager
     ***************************************************************************/
 
-    private void setNextRevealHeight (Height height) @safe nothrow
+    private size_t loadIntegerData (string name, size_t default_value = 0)
```
If you change it to `T loadIntegerData (T)(string name)` you won't have to do any casting. The same goes for `saveIntegerData`, writing `void saveIntegerData (T)(string name, T value)`.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -659,53 +693,77 @@ public class EnrollmentManager
+            if (results.empty)
+                return default_value;
+
+            return results.oneValue!(size_t);
```
And then this becomes `oneValue!T` after the change.

linked0 commented:
Applied the templates. Thanks.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1142,3 +1158,62 @@ unittest
+    auto gen_key_pair = WK.Keys.Genesis;
+    auto key_pair = KeyPair.random();
+
+    foreach (uint idx; 0 .. 7)
+    {
+        auto tx = TxBuilder(GenesisTransaction, idx).
+            draw(Amount(Amount.MinFreezeAmount), [gen_key_pair.address]).
+            sign(TxType.Freeze);
+        utxo_set.put(tx);
+    }
```
This is invalid, the UTXOs will double spend, which only works because we don't have a ledger instance.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1142,3 +1158,62 @@ unittest
+
+    // create enrollments for nodes
+    auto params = new immutable(ConsensusParams)();
+    auto man = new EnrollmentManager(":memory:", gen_key_pair, params);
```
We should stop using the genesis keypair for those tests, it's a weird edge case that is no realistic.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1142,3 +1158,62 @@ unittest
+    PreImageInfo org_preimage;
+    assert(man.getPreimage(Height(10), org_preimage));
+
+    // clear up the EnrollmentManager for test
+    man = new EnrollmentManager(":memory:", gen_key_pair, params);
+    assert(man.addValidator(enrolls[0], Height(2), finder, utxos) is null);
+    assert(man.addValidator(enrolls[1], Height(5), finder, utxos) is null);
+    assert(man.addValidator(enrolls[2], Height(7), finder, utxos) is null);
+
+    // check the restored enrollment data
+    assert(man.enroll_key == enrolls[2].utxo_key);
+    assert(man.data == enrolls[2]);
+
+    // check the restored pre-image
+    PreImageInfo preimage;
+    assert(man.getPreimage(Height(10), preimage));
+    assert(preimage == org_preimage);
```
Yeah this test doesn't actually test much.

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +143,81 @@ unittest
     assert(b20.header.enrollments[0] == enroll);
 }
+
+/// Situation: A node in the network crashes after a new enrollment for the
+///     node has been inserted into a new block with consensus. And the node
+///     starts again
+/// Expectation: The node restores its own enrollment from the chain and
+///     reveals its pre-images periodically.
```
👍

Geod24 commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -142,3 +143,81 @@ unittest
+///     reveals its pre-images periodically.
+unittest
+{
+    // Boilerplate
+    const validator_cycle = 20;
+    TestConf conf = {
+        validator_cycle : validator_cycle,
+    };
```
Not needed anymore

linked0 committed:
Bug fix: prevent not getting a pre-image

If the formula of `next_reveal_height + PreimageRevealPeriod * 2` in
the `getNextPreimage` function returns a value that exceeds the
validator cycle, the function sets the height of the pre-image to
ulong.max. So, the `getNextPreimage` can return the last pre-image
if a node is a validator.

linked0 committed:
Save the cycle index of pre-images

When the `addValidator` is called, the cycle count for enrollments should
be restored. First, we add new column named 'cycle_count` into the table, and
when the `addValidator` function is called, the `cycle_count` is saved. After
that, when the `EnrollmentManager` restarts, it loads the count and calls the
`populate` function as times as the count.

linked0 committed:
Bug fix: reset enrollment after the cycle ends

There was a bug that the `getPreimage` function is called even if
the validator cycle of a node ends. It is because the enrollment
was not reset after the validator cycle ends. So in this commit,
I clear up the enrollment of the node.

linked0 committed:
Restore the enrollment data from the Genesis block

There should be one point where we can add validators outside the
`ValidatorSet` but the `restoreValidators` of the `EnrollmentManager` is
another path to adding validators which could produce bugs. This makes
the enrollment of a node like the cycle index not restored form
the GenesisBlock.

linked0 commented:
Ready for review.
It might look that these commits are not related with each other, but in fact, there is relativeness between them. In overall, the commits are for making the pre-images revealed normally after restoring validators from the Genesis block.
I have tried to make a network test but there is some obstacle for that. I would be happy to have pair-programming for making a network test if possible. Before that, I will do more analysis. I already checked the cycle index is increased after restoring validators from the Genesis block by printing the log in the existing network test. Is it possible to have another PR for the network test? Thanks.

linked0 commented:
I will add a network test for this PR and request a review again.

linked0 commented:
I added a network test. By the way, there was a critical bug about revealing pre-images of the enrollments of the Genesis block. The last commit and the network test is for the bug. Before the completion of the CI builds, you can review my solution for that. I hope it's the proper workaround. I will have a white-boarding session on how to resolve the root cause of this bug next week if needed.

linked0 committed:
Bug fix: Make the validators of Genesis reveal pre-images timely

I fixed using the wrong type (ulong instead of uint) when creating
a pre-image hash. The bug caused the failure of revelation of pre-images of
the validators of the Genesis block. And I added a network test for restoring
the enrollment information of a node and revealing pre-images of `Enrollment`s
of the Genesis block.

linked0 commented:
Ready for review.
I found out and applied the right solution for the bug related to revealing pre-images of the enrollments of the Genesis block. Besides the bug, this PR has many commits fixing other bugs. Thanks.

AndrejMitrovic commented on source/agora/consensus/validation/PreImage.d:
```diff
@@ -53,8 +53,8 @@ public string isInvalidReason (const ref PreImageInfo new_image,
-    if (new_image.distance < prev_image.distance)
-        return "The height of new pre-image is smaller than that of previous one";
+    if (new_image.distance <= prev_image.distance)
+        return "The height of new pre-image is equal to or smaller than that of previous one";
```
```suggestion
        return "The height of new pre-image is not greater than that of the previous one";
```

AndrejMitrovic replied:
Where is the associated unittest for this change?

Geod24 replied:
Moved to #1040

AndrejMitrovic commented on source/agora/test/EnrollmentManager.d:
```diff
@@ -221,3 +221,32 @@ unittest
+    };
+    auto network = makeTestNetwork(conf);
+    scope(exit) network.shutdown();
+    // scope(failure) network.printLogs();
```
```suggestion
    scope(failure) network.printLogs();
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -390,7 +390,7 @@ public class EnrollmentManager
         auto cache = PreImageCache(PreImageCycle.NumberOfCycles, cycle_length);
         assert(offset < cache.length);
-        cache.reset(hashMulti(kp.v, "consensus.preimages", offset));
+        cache.reset(hashMulti(kp.v, "consensus.preimages", cast(uint)offset));
```
The commit message should say what the actual bug was. For example:

```
Fixed using the wrong type (ulong instead of uint) when creating the preimage hash.
```

Geod24 replied:
We should just change the parameter type maybe

linked0 commented:
@AndrejMitrovic  Thank you for your review. I will apply the comments to the code.

Geod24 commented:
Can you extract afa15052b4da564505cd9d2693162b77ea0a9606 into its own PR, plus the associated network test ?
Thanks.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -470,7 +470,12 @@ public class EnrollmentManager
+        if (enrolled == ulong.max)
+            return false;
+
+        auto height = Height(min(this.next_reveal_height + PreimageRevealPeriod * 2,
```
`const  height`

Geod24 replied:
Regarding the commit message:

> If the formul**a** [...]
> [...] returns ~tha~ **a** value [...]
> [...] the function set*s* the height [...]
> [..] can return~s~ the last [...]

linked0 replied:
> Regarding the commit message:
> 
> 
> 
> > If the formul**a** [...]
> 
> > [...] returns ~tha~ **a** value [...]
> 
> > [...] the function set*s* the height [...]
> 
> > [..] can return~s~ the last [...]

I'll find a plugin for spell-checker for (neo)vim for sure.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -289,6 +294,10 @@ public class EnrollmentManager
+
+            // save the count of 'populate' of `PreImageCycle`
+            this.setCycleIndex(
+                this.cycle.nonce * this.cycle.NumberOfCycles + this.cycle.index);
```
Could you parenthesize these? `(this.cycle.nonce * this.cycle.NumberOfCycles) + this.cycle.index` or whichever precedence it is (it is not obvious).

linked0 replied:
> Could you parenthesize these? (this.cycle.nonce * this.cycle.NumberOfCycles) + this.cycle.index or whichever precedence it is (it is not obvious).

Your suggestion could be more readable.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -751,6 +760,55 @@ public class EnrollmentManager
+    {
+        try
+        {
+            auto results = this.db.execute("SELECT val " ~
+                "FROM node_enroll_data where key = ?", "cycle_index");
```
Why not just this:

```suggestion
            auto results = this.db.execute("SELECT val " ~
                "FROM node_enroll_data where key = cycle_index");
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -751,6 +760,55 @@ public class EnrollmentManager
+        catch (Exception ex)
+        {
+            log.error("ManagedDatabase operation error {}", ex);
+            return 0;
```
Will the calling code properly handle this condition?

linked0 replied:
It's a proper question. But, It's a difficult problem. If there's a problem with restoring the cycle index, what should we do? The only way to restore the cycle index is to search for all the enrollment from the Genesis block to the lastest block. We can find a right solution for that if we research on the issue #849  I guess. Anyway, I have two ideas about what you are questioning. What do you think is better for that at the moment between just adding `assert(0)` or restoring it from looking into all the blocks?

AndrejMitrovic replied:
We need to get some better error information from sqlite before we can do proper error handling (see for example https://github.com/bpfkorea/d2sqlite3/pull/1).

For now we could proceed with this try/catch code. Later we will have to fix our usage of sqlite.

linked0 replied:
> We need to get some better error information from sqlite before we can do proper error handling (see for example bpfkorea/d2sqlite3#1).

I'll take a look at the PR. Thanks.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -474,6 +474,16 @@ public class EnrollmentManager
+        if (block_height >= enrolled + params.ValidatorCycle)
+        {
+            this.enroll_key = Hash.init;
+            this.data = Enrollment.init;
```
What about `signature_noise` and `cycle`? Do they also need to be reset? It might be useful to move this into a new `reset()` function, or named something else more appropriate.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -474,6 +474,16 @@ public class EnrollmentManager
     {
+        // clear up the enrollment of a node if the validator cycle of the node
+        // ends at the `block_height`
+        const enrolled = this.validator_set.getEnrolledHeight(this.enroll_key);
```
It might be useful to move this into a private function called `isEnrolled()`. Then inside of `getPreimage` you could add an assert: `assert(!this.isEnrolled());` to catch programming bugs.
